### ISSUE TITLE
High resource consumption when using large bundles

### ISSUE THREAD

**Issue by 0xAP** [open-policy-agent/opa#3860]
I'm currently evaluating OPA's performance in different setups and policy/data sizes. The decission latencies as well as the flexiblity make it a great technology. However, for one of our use case we observed some issues. 

**Context**
For some use cases we want to utilize the bundle mechanism to distribute both policies and data. We want to do that in order to allow low-latency decission making while having a higher level of resiliance against network outages or server problems.

**Observation - long startup times**

With an increase of the uncompressed bundle size the startup time of opa agent increases more than linearly.
```
opa run -b bundle_100MB.tar.gz --server // 100MB uncompressed
opa run -b bundle_200MB.tar.gz --server // 200MB uncompressed
```
Startup times:
* 100MB -> took 16s
* 200MB -> took 71s
* 500MB -> took 690s

The bundles had the structure:
```
 xxxxx1.rego
 xxxxx1.json
	...

 xxxxx<COUNTER>.rego
 xxxxx<COUNTER>.json
```
Where each file tuple (ie. `xxxxxN.rego` and `xxxxxN.json`) had total uncompressed size of 1000 Bytes.

The files look as follows:

**Rego file**
```
package benchmark.<COUNTER> 
default allow = false
allow {
  data.benchmark0<COUNTER>.a.b.c.d.key = input.key
  data.benchmark1<COUNTER>.e.f.g.h.id = input.id
  data.benchmark2<COUNTER>.i.j.k.l.name = input.name
}
```
**Data file**
```
{
  "bytesfiller<COUNTER>": "123",
  "benchmark0<COUNTER>": {
    "a": {
      "b": {
        "c": {
          "d": {
            "key":"fookey"
          }
        }
      }
    }
  },
  ...
}

```
In case you are wondering about `<COUNTER>`: in order to generate a bundle of size X we are generating multiple of those tuples by `<COUNTER>` is replaced with an increasing integer value.

**Problem:**

For attribute-rich use-cases that require a low latency for decission making this loading times are problematic, since instances take a very long time till they are up and ready, hence, making scaling challenging. Also a lot of resources are used (I am aware that OPA data structures are about 6x the size of the raw json).

**Findings: heavy memory allocations and CPU usage**

We investigated both total memory allocation as well as CPU consumption using golang's pprof (we are only going to talk about the memory since the CPU pretty much tells the same story)

```
opa run -b bundle_<XYZ>MB.tar.gz --server // pprof enabled

go tool pprof http://localhost:6060/debug/pprof/allocs

```
The total cumulated allocations were:
* 100MB -> ~10GB 
* 200MB -> ~41GB 

![Screenshot 2021-10-05 at 10 12 57](https://user-images.githubusercontent.com/22645579/135985609-94883393-a7f4-44b3-8c2f-1c917b82172f.png)
![Screenshot 2021-10-05 at 10 13 04](https://user-images.githubusercontent.com/22645579/135985768-bb7a8986-e516-46ed-b437-6185e0292c80.png)

More than 95% of all allocations are coming from `/internal/merge.InterfaceMaps`. This function is recursively called and with each call the current global data object is copied over to a new object.
We assume that this design was purposely choosen to enforce immutability of the global data store. However, checking the references of `merge.InterfaceMaps` we found that it is only used in contexts where immutability is not an issue.

We applied a patch to `merge.InterfaceMaps` which realizes the merge in a mutable fashion. Benchmarking the startup time of a 500MB file we found that it could be reduced from `690s` to `17s`.

```
func InterfaceMaps(a map[string]interface{}, b map[string]interface{}) (c map[string]interface{}, ok bool) {

   // this causes the high allocations
	c = map[string]interface{}{}
	for k := range a {
		c[k] = a[k]
	}
```

# Questions
* without talking about specific implemenation yet, would it in general be possible to make the `InterfaceMaps` a mutable operation or would it cause problems in other spots?

**Comment by srenatus** [comment]
Thanks for digging into this and sharing your findings :+1:

Given that it's an internal package, I'd suppose we could change its inner workings if the effect is so dramatic.... That said, I haven't reviewed the call sites yet. Would changing this still pass all tests? I guess if you opened a PR with your patch, we could better discuss this... 👀

**Comment by anderseknert** [comment]
Yeah, definitely interesting! It was I who asked for an issue before a PR, so that we’d have an idea what we were dealing with @srenatus :) Either way good to have an issue to connect said PR to later. But I agree, PR to see if/how things are affected would be great at this point.

**Comment by 0xAP** [comment]
Thanks for the quick replies. @srenatus yes the tests were still passing with our patch. I'll do a PR and link this issue to it.

**Comment by Sulbigar** [comment]
This is good research. Thank You @0xAP. I was planning to do similar profiling exercise as we recently started doing our POC recently. Will be following this thread.

**Comment by thilak009** [comment]
> I am aware that OPA data structures are about 6x the size of the raw json

@0xAP where did you find this information? i have been trying to read and understand this and wasn't able to find anything in the documentation 
if you could point me to any resources, that would be helpful, thank you
